### PR TITLE
promote Mixup and Cutmix from prototype to transforms v2

### PR DESCRIPTION
**Pull request by pmeier**
TL;DR: this PR promotes Mixup and Cutmix from `torchvision.prototype.transforms` to `torchvision.transforms.v2` by reusing the `labels_getter` functionality that we have for `SanitizeBoundingBoxes`.

To achieve this, the following this are implemented here:

1. Factor out the static "`labels_getter`" methods from `SanitizeBoundingBoxes`. While doing that we also change the functionality slightly to make the handling a little easier:
   - Remove the functionality of passing a string to `labels_getter`, since that is just slightly more convenient than passing a callable directly, while making the handling harder for us.
   - Add the functionality to return labels if we find a tensor as second element of a tuple. We need this for Mixup and Cutmix.
2. Remove the `p` parameter. We have this parameter in our references, since we have based them on a research implementation:
   https://github.com/pytorch/vision/blob/08c9938f3e69b4fb2d710a038479264f3c23b133/references/classification/transforms.py#L22
   However, by design, a research implementation will have more knobs than a stable library. In fact, we are hardcoding the parameter in our references:
   https://github.com/pytorch/vision/blob/08c9938f3e69b4fb2d710a038479264f3c23b133/references/classification/train.py#L224-L227
   By removing the `p` parameter in this PR, we get the same behavior that we currently have in our references as well. If the user need the more flexible behavior back, they can always wrap the transform like `RandomApply(Mixup(...), p=...)`. 

   Since the existence of the `p` parameter was the reason to prefix "Random" before the "canonical" names Mixup and Cutmix, I've dropped the prefix here as well.
3. Follow-up to #7724. The implementation of this PR will be available in the classification references with `--use-v2`. I also refactored the training script to just use the transform inside the training loop rather than putting it inside the `collate_fn`. This makes it more clear that this transform needs batching, but is otherwise independent of the data loader.

### ToDo

- [ ] Fix old tests for `SanitizeBoundingBoxes`
- [ ] Write new tests for `Mixup` and `cutmix`
- [ ] Remove prototype transforms `RandomMixup` and `RandomCutmix`

cc @vfdev-5

### CODE REVIEW

File: references/classification/train.py
Comment (NicolasHug) [inline references/classification/train.py:29]: I failed to notice this when we discussed it offline, but we should keep those transforms as `collate_fn`: calling them after the dataloder like done here means we can't leverage multi-processing.

File: references/classification/train.py
Comment (NicolasHug) [inline references/classification/train.py:15]: Sooooo to avoid bikeshedding on how we should call those (batch transforms vs pairwise transforms vs something else), maybe we should just rename that to `get_cutmix_mixup`?

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:166]: Maybe we can write that entire message regardless of whether "default" was passed. It would simplify the logic a bit and avoid storing `self.labels_getter`.

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:167]: Technically this could qualify as a ValueError as well?

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:173]: What does it mean to be probability based? We probably just mean "one-hot-encoded" for the input (ground-truth) labels right?

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:180]: minor typos: remove "a" in both lines (and clarify what "probability-based" means)

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:193]: end of comment missing :)

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:194]: We used a different strategy in `SanitizeBoundingBox` where we called `_transform()` on all inputs and just handled that filtering logic within `_transform()`. I don't have a pref right now (haven't thought about it much). But maybe we should align both transforms to follow the same strat? (we could do it in another PR)

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_augment.py:217]: an -> a

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip, only made minor comments for now, overall this looks great. Will give a more thorough look once we have the tests

File: torchvision/transforms/v2/_augment.py
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:173]: We had some major discussions about this before you joined. The problem with saying "one-hot encoded labels" is that after Cutmix or Mixup that is no longer true. One-hot encoding basically means turning `3` into `[0, 0, 0, 1, 0]` given we have 4 categories. However, due to the nature of Mixup and Cutmix, afterwards the label could look like `[0, 0.5, 0, 0.5, 0]` if an image with label `1` was mixed with one of label `3`. And that can no longer be called "one-hot encoding".

To find better suiting terms, I looked at some loss functions, e.g. [`CrossEntropyLoss`](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html?highlight=crossentropyloss#torch.nn.CrossEntropyLoss):

> Target: If containing **class indices**, shape ()(), (N)(N) or (N,d1,d2,...,dK)(N,d1​,d2​,...,dK​) with K≥1K≥1 in the case of K-dimensional loss where each value should be between [0,C)[0,C). If containing **class probabilities**, same shape as the input and each value should be between [0,1][0,1].

(emphasis mine)

And this is why I'm using the terms "index based" and "category based" here.

File: torchvision/transforms/v2/_augment.py
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:194]: We can't use the same strategy here. `SanitizeBoundingBox` does not affect images or videos, so we don't care about `needs_transform_list` there. However, here we transform images. Meaning, we need to use `needs_transform_list` to make use of the heuristic about what image to transform. This cannot be done in `_transform` since in there we have no concept if an image should be transformed or not.

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_augment.py:173]: I understand the ambiguity for labels *after* they're transformed by CutMix/MixUp, but when we're talking about the input (i.e. before being transformed), the labels tensor should always be one-hot-encoded (or just ints) in the stricter sense, right?

Since this error message is only referring to the input, I feel like we can go ahead and use "one-hot-encoded" safely?

File: torchvision/transforms/v2/_augment.py
Comment (vfdev-5) [inline torchvision/transforms/v2/_augment.py:157]: Do we really need this `needs_transform_list` ?
We can know where labels are and we pass through non-handled types, it feels like we can just transforms all items of the input ?

Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:157]: See https://github.com/pytorch/vision/pull/7731#discussion_r1259519124

File: torchvision/transforms/v2/_augment.py
Comment (vfdev-5) [reply torchvision/transforms/v2/_augment.py:194]: Shouldn't we transform all images ?

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_misc.py:304]: we'll want to update the docstring above I think

File: torchvision/transforms/v2/_utils.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_utils.py:126]:
```suggestion
            f"When using the default labels_getter, the input passed to forward must be a dictionary or a two-tuple "
```

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_augment.py:194]: I think I understand what you mean: we'd need to re-implement the "tensor pass-through heuristic" in `_transform()` if we were to do something like in `SanitizeBoundingBox()`, and we don't want to do that. I feel like we could use the same strategy used here in `SanitizeBoudingBox()` though. But that's OK.

> Shouldn't we transform all images (each image is collated as (N, C, H, W)) ?

We are transforming all images yes

File: torchvision/transforms/v2/_utils.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_utils.py:101]: Good point' it'd be interesting to figure out whether this makes things faster. Might be best to leave this out as future improvement though, to keep this PR simpler.
One thing to note: doing this would tie the transform instance to a specific dataset [format]. IDK whether this is a problem in practice, but worth keeping in mind

File: torchvision/transforms/v2/_augment.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_augment.py]: > the labels tensor should always be one-hot-encoded

I guess that's not the case, typically when one combines `Compose([CutMix(), MixUp()])`. The transforms seem to work fine even when "dirichlet-encoded" inputs are passed in, which seem to be by design?

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1688]: Sooo I'm writing this because this is what I observed. But in practice I' not sure when this happens and whether this is expected. Is it because the `Beta` distribution might end up on a value of `1` for `lam`?

CC @pmeier

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1643]: Yes it does and exactly for the use case you put there.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1644]: Instead of parametrizing over the class / name here, should we maybe parametrize over the instance along with `expected_num_non_zero_labels`? Plus, we should probably also do the "reversed compose", i.e. Mixup before Cutmix.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1652]: Do we actually need this? If we go for the "reverse compose" as suggested above, each transform automatically gets passed probability based labels from the output of the transform that comes before.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1687]: I mean, it is possible, but vanishingly unlikely. For the default value of `alpha=0.5`, the values concentrate around 0 and 1, but hitting a value that is close enough to vanish inside the floating point precision is rather unlikely.

If we want to make sure, we can always fix the random seed or mock the distribution to return `0.5` or the like.

Lastly, we could also set `alpha=10` or the like, which heavily concentrates distribution around `0.5` and thus again making the no-op case highly unlikely.

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1705]: Do we need to have this is "separate" test or can we just do

```python
img, target = packed_from_dl = next(iter(dl))
```

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1742]: ?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:1750]: Since all errors are runtime errors below, can we just do `tranform = T(alpha=0.5)` upfront rather than instantiating it every time?

File: test/test_transforms_v2_refactored.py
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1705]: Whether both cases should be supported, we can discuss. But right now we're testing both (as they're different). I can add the line you suggested but it's saving one line. Is it necessary?

File: test/test_transforms_v2_refactored.py
Comment (pmeier) [reply test/test_transforms_v2_refactored.py]: Sorry, I misplaced the comment. It should have been on the extra `ToOneHot` transform that we add below. If we have all four variants (both transforms individually and composed in both orders) we are sure that each one get a probably-based label at some point. Thus, we can drop the `one_hot` parametrization.

Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py]: As discussed offline I'm going to remove testing for Compose because this is probably mis-using the transforms. So we still need to have this `one_hot` parametrization

**Comment by NicolasHug** [comment]
Quick update: after chatting with @pmeier we decided to remove support for 2D labels. There isn't a strong need for it considering CutMix and MixUp should probably never be called consecutively - it's either one of the other.

Should users request this to be supported for whatever reason (maybe Compose(CutMix(), CutMix()) make sense??) then we can add this back.

support was removed in [9f4a9e6](https://github.com/pytorch/vision/pull/7731/commits/9f4a9e607d217c8f4cc2f02084006d1db787d895)

File: torchvision/transforms/v2/_misc.py
Comment (pmeier) [reply torchvision/transforms/v2/_misc.py:304]: Ping

File: torchvision/transforms/v2/_utils.py
Comment (pmeier) [inline torchvision/transforms/v2/_utils.py:117]: Flag

File: torchvision/transforms/v2/_utils.py
Comment (pmeier) [inline torchvision/transforms/v2/_utils.py:131]: Nit: can we remove this comment now that we have detailed description in the docstring?

File: torchvision/transforms/v2/_utils.py
Comment (NicolasHug) [reply torchvision/transforms/v2/_utils.py:117]: Sooo I've decided not to document it in the code because this will probably just add more confusion. But the reason we need to add support for `list` is because this is what the DataLoader actually returns (in its most default setting):

```py
for x in DataLoader(...):
    # x is  a list [img_batch, labels_batch] and we want to support
    CutMix(x)
```

File: torchvision/transforms/v2/_misc.py
Comment (NicolasHug) [inline torchvision/transforms/v2/_misc.py:279]: I chose not to document the "labels" key matching in the finer details. We can revisit (or point users to check the code?)

File: gallery/plot_cutmix_mixup.py
Comment (NicolasHug) [inline gallery/plot_cutmix_mixup.py:7]: https://github.com/pytorch/vision/issues/7766

File: references/classification/train.py
Comment (NicolasHug) [inline references/classification/train.py:16]: I'll validate the changes made to the references once this is merged.

File: test/test_transforms_v2.py
Comment (NicolasHug) [inline test/test_transforms_v2.py:1740]: Note that I had to delete this. I feel like it was a valid error to raise. We could put it back if we were to have a different `labels_getter` logic for SanitizeBBox and the Cutmix/Mixup ones (which is probably going to be needed eventually anyway).

This is OK for now.

**Review by NicolasHug — APPROVED** [review]
I think this LGTM. Since I made a bunch of changes, this should get another round of reviews from @pmeier @vfdev-5 before merging

File: torchvision/transforms/v2/_augment.py
Comment (vfdev-5) [reply torchvision/transforms/v2/_augment.py:156]: Linked comment does not respond to my comment. As we talked previously, the way how it is coded here, we do not need to create redundant `needs_transform_list` structure and can pass all flat inputs items into `_get_params` and `_transform`. Those functions are handling unsupported types internally.

**Comment by vfdev-5** [comment]
There are conflicts to resolve

**Comment by NicolasHug** [comment]
They're trivial, I'll address at the next (and hopefully last) review

File: torchvision/transforms/v2/_augment.py
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:156]: As discussed offline, this is needed. Imagine a sample like

```python
sample = {
    "image": datapoints.Image(...),
    "label": torch.Tensor(...),
    "some_other_thing": torch.Tensor(...),
}
```

Without `needs_transform_list`, we cannot prevent `"some_other_thing"` to be transformed as an image.

File: torchvision/transforms/v2/_augment.py
Comment (pmeier) [reply torchvision/transforms/v2/_augment.py:179]: > I feel like we could use the same strategy used here in `SanitizeBoudingBox()` though. But that's OK.

Yes, we can certainly also use `needs_transform_list` there. I'm ok with that. Up to you.

**Review by pmeier — COMMENTED** [review]
LGTM, thanks Nicolas for finishing this. I can't approve though, since it is technically my PR.

File: docs/source/transforms.rst
Comment (vfdev-5) [inline docs/source/transforms.rst:264]: Nit: technically, paper names of these techniques are  CutMix and MixUp.

Comment (NicolasHug) [reply docs/source/transforms.rst:264]: Sorry, saw this after I merged. I'll address via https://github.com/pytorch/vision/issues/7766

**Pull request merged**
